client/network/src/request_responses.rs

```diff
@@ -738,23 +630,58 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
 							self.pending_responses_arrival_time
 								.insert((protocol.clone(), request_id).into(), Instant::now());
 
-							let get_peer_reputation = self.peerset.clone().peer_reputation(peer);
-							let get_peer_reputation = Box::pin(get_peer_reputation);
+							let reputation = self.peer_store.peer_reputation(&peer);
 
-							// Save the Future-like state with params to poll `get_peer_reputation`
-							// and to continue processing the request once we get the reputation of
-							// the peer.
-							self.message_request = Some(MessageRequest {
-								peer,
-								request_id,
-								request,
-								channel,
-								protocol: protocol.clone(),
-								resp_builder: resp_builder.clone(),
-								get_peer_reputation,
-							});
+							if reputation < BANNED_THRESHOLD {
+								log::debug!(
+									target: "sub-libp2p",
+									"Cannot handle requests from a node with a low reputation {}: {}",
+									peer,
+									reputation,
+								);
+								continue 'poll_protocol
+							}
+
+							let (tx, rx) = oneshot::channel();
+
+							// Submit the request to the "response builder" passed by the user at
+							// initialization.
+							if let Some(resp_builder) = resp_builder {
+								// If the response builder is too busy, silently drop `tx`. This
+								// will be reported by the corresponding request-response
+								// [`Behaviour`] through an `InboundFailure::Omission` event.
+								// Note that we use `async_channel::bounded` and not `mpsc::channel`
+								// because the latter allocates an extra slot for every cloned
+								// sender.
+								let _ = resp_builder.try_send(IncomingRequest {
```

Comment:
There is still one problem in this code. Syncing only accepts 40 peers and processes received block announcements from those peers and discards others but the request-response code accepts requests from any connected peer. These connected but rejected peers are allowed to consume the request-response channel slots which is not very optimal as you can imagine. I'm not sure how to fix this though. We need take a look at this code again in the near future.

Reply:
Not sure how to implement it now, but may be request-response can subscribe to connection events from block announces notifications protocol and ensure only the connected peers are served.